### ISSUE TITLE
Next planned charging period?

### ISSUE THREAD

**Issue by gcoan** [springfall2008/batpred#416]
I thought it would be useful to put on my dashboard what the current period rate and when predbat is next due to charge and discharge and at what rates.

current period rate is predbat.rates and predbat.rates_export

But next charge and discharge seem to be harder to find.

I've got next low rate start, end and cost and same for high rate:
![image](https://github.com/springfall2008/batpred/assets/142018870/152483a8-acb8-41f4-883b-98a96b2d6253)

But looking at the plan, 20:00-20:30 is not planned for import nor 20:00-21:00 for export (and in fact the next best export period would be tomorrow's peak):
![image](https://github.com/springfall2008/batpred/assets/142018870/de8b7273-0eba-44a5-90f4-d7e51923a120)
and then a minute later the 20:00 slot became freeze discharge (soc is currently 4% so its a moot point)
![image](https://github.com/springfall2008/batpred/assets/142018870/38a1b8be-79d2-4006-9eaf-3ff9e471ef79)
And then at 20:20 the latest plan swapped back from 20:00 being a freeze discharging slot!

I looked at the attributes of predbat status and the best_charge_window is encoded there, but not sure what the start and end refer to - minutes from now?
![image](https://github.com/springfall2008/batpred/assets/142018870/3de4385c-cc24-4246-9a06-309ae9190b11)

Is this information easily accessible?  Are the next low and high rates being populated correctly?

**Comment by springfall2008** [comment]
I think the problem is for Agile the next low and high rate slots aren't very meaningful as most slots are considered but many aren't used.

Do you want the next charge slot and the next force export slot or something else?

**Comment by gcoan** [comment]
Agreed. I’d like to know the next charge slot and next force export slot, although the export is far lesser priority as very little force exporting at the moment and I know when the higher rate export slots are.

Thanks

**Comment by gcoan** [comment]
Thanks @springfall2008, so making sure I understand the results attribute correctly, its a list of charging limits and start times.  So the next charge slot will be the first array entry in results[] where the % is >0,  which will be results[0] if we're currently in a charge slot, or results[1] if its in the future?
![image](https://github.com/springfall2008/batpred/assets/142018870/f458ceae-fbcf-4edf-a2f9-56d8434a6a6a)

**Comment by springfall2008** [comment]
I’d ignore the results part it’s for the chart.

There’s a start and end time sensor and then the rate attribute gives you the import rate

**Comment by gcoan** [comment]
I upgraded to the main branch.  Can see predicted charge start time sensors now, don't know if these are new or I Just never saw them before?

Curiously there seems to be gaps in the charge start time sensor and its currently undefined:
![image](https://github.com/springfall2008/batpred/assets/142018870/3793d82d-ba82-4261-b9da-19b00dc86a91)

![image](https://github.com/springfall2008/batpred/assets/142018870/a731ec7a-348a-4704-b580-df972c311314)

But the plan clearly shows charging in the octopus power up event this afternoon:
![image](https://github.com/springfall2008/batpred/assets/142018870/74edd02d-3877-4cc3-bf28-9b8d3ff24923)

Here's the logfile
[appdaemon (1).log](https://github.com/springfall2008/batpred/files/13581801/appdaemon.1.log)

**Comment by springfall2008** [comment]
Are you looking at the best version of these sensors ?

**Comment by springfall2008** [comment]
Start and end time best are what’s planned. The other ones are what’s configured currently

**Comment by gcoan** [comment]
OK found them:
          - entity: predbat.best_charge_start
          - entity: predbat.best_charge_end
          - entity: predbat.best_discharge_start
          - entity: predbat.best_discharge_end
Thank you

**Comment by springfall2008** [comment]
Yes sorry, and limit has the rate encoded
